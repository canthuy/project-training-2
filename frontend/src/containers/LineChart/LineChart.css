.data-group {
  display: flex;
  align-items: center;
  justify-content: space-between;
}
.data-group span {
  font-size: 16px;
  font-weight: 600;
  color: #111;
  cursor: pointer;
  margin: 0 10px;
}
