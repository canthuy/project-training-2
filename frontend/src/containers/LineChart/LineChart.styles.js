import styled from 'styled-components';

const ChartSC = styled.div`
  width: 70%;
  margin: 20px;
`;

const Title = styled.h5`
  margin: 0;
  margin-left: 10px;
  color: #7a7975;
`;
export { ChartSC, Title };
