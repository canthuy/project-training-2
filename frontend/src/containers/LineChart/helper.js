import moment from 'moment';

const handleData = (deviceData, formatType) => {
  const newSeries = deviceData.map((device) => {
    const data = device.data.map((day) => ({
      x: moment(day.x, 'DD/MM/YYYY').format(formatType),
      y: day.y,
    }));
    let obj = {};
    const newData = data.reduce((totalY, day) => {
      if (!obj[day.x]) {
        obj[day.x] = day;
        totalY.push(obj[day.x]);
      } else {
        obj[day.x].y += day.y;
      }
      return totalY;
    }, []);
    return {
      name: device.name,
      data: newData,
    };
  });
  return newSeries;
};

export { handleData };
