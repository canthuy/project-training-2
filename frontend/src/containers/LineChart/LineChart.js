import { useSelector, useDispatch } from 'react-redux';
import { Fragment, useCallback, useEffect, useMemo, useState } from 'react';
import Chart from 'react-apexcharts';
import Loading from '../../components/Loading/Loading';
import Error from '../../components/Error/Error';
import { ChartSC, Title } from './LineChart.styles';
import { getData } from '../../redux/lineChart/actions';
import { handleData } from './helper';
import './LineChart.css';

const LineChart = () => {
  const { deviceData, loading, isError } = useSelector(
    (state) => state.lineChart
  );
  const { startDate, endDate } = useSelector((state) => state.datepicker);
  const dispatch = useDispatch();
  const groupData = ['Day', 'Week', 'Month'];
  const [isActive, setIsActive] = useState(0);
  const [series, setSeries] = useState([]);

  useEffect(() => {
    dispatch(getData([startDate, endDate]));
  }, [startDate, endDate]);
  useEffect(() => {
    setSeries(deviceData);
    setIsActive(0);
  }, [deviceData]);
  const totalData = useMemo(() => {
    const total = deviceData.map((val) =>
      val.data.reduce((acc, val) => acc + val.y, 0)
    );
    return total;
  }, [deviceData]);
  const options = {
    chart: {
      toolbar: {
        show: false,
      },
      animations: {
        speed: 1000,
        easing: 'easein',
      },
    },
    legend: {
      show: true,
      horizontalAlign: 'left',
      markers: {
        width: 16,
        height: 10,
        radius: 4,
      },
      itemMargin: {
        horizontal: 10,
      },
      formatter: (seriesName, opts) =>
        seriesName + ' ' + totalData[opts.seriesIndex],
    },
    stroke: {
      width: 4,
    },
    colors: ['#925de2', '#48c0b0'],
  };
  const handleClick = useCallback(
    (index) => {
      setIsActive(index);
      if (index === 0) {
        setSeries(deviceData);
      }
      if (index === 1) {
        const newSeries = handleData(deviceData, 'WW');
        setSeries(newSeries);
      }
      if (index === 2) {
        const newSeries = handleData(deviceData, 'MM/YYYY');
        setSeries(newSeries);
      }
    },
    [deviceData]
  );
  return (
    <>
      <ChartSC>
        {loading && <Loading />}
        {isError && <Error message="Error: Network Error" />}
        {!loading && !isError && (
          <Fragment>
            <div className="data-group">
              <Title>Device</Title>
              <div>
                {groupData.map((val, index) => {
                  return (
                    <span
                      key={index}
                      style={{ opacity: isActive === index ? 1 : 0.3 }}
                      onClick={() => handleClick(index)}
                    >
                      {val}
                    </span>
                  );
                })}
              </div>
            </div>
            <Chart options={options} series={series} type="line" height={404} />
          </Fragment>
        )}
      </ChartSC>
    </>
  );
};

export default LineChart;
