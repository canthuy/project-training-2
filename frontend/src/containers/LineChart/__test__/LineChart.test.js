import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { useSelector, useDispatch } from 'react-redux';
import LineChart from '../LineChart';
window.React = React;

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));
jest.mock('react-apexcharts', () => 'Chart');
describe('LineChart component', () => {
  let mockState = {};
  beforeEach(() => {
    useDispatch.mockImplementation(() => () => { });
    useSelector.mockImplementation((callback) => {
      return callback(mockState);
    });
  });
  afterEach(() => {
    useSelector.mockClear();
    useDispatch.mockClear();
  });
  it('loading data', () => {
    mockState = {
      lineChart: {
        deviceData: [],
        loading: true,
        isError: false,
      },
      datepicker: {
        startDate: '01/01/2021',
        endDate: '01/02/2021',
      },
    };
    const { container } = render(<LineChart />);
    const elements = container.querySelector('div').firstChild;
    expect(container).toMatchSnapshot();
    expect(elements.firstChild.className).toEqual('spinner-border');
  });
  it('get data success', () => {
    mockState = {
      lineChart: {
        deviceData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 4 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
            ],
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 1 },
              { x: '02/01/2021', y: 6 },
              { x: '03/01/2021', y: 4 },
            ],
          },
        ],
        loading: false,
        isError: false,
      },
      datepicker: {
        startDate: '01/01/2021',
        endDate: '01/02/2021',
      },
    };
    const { container, getByText } = render(<LineChart />);
    expect(container).toMatchSnapshot();
    expect(getByText(/Device/i)).toBeTruthy();
  });
  it('get data success and group by day', () => {
    mockState = {
      lineChart: {
        deviceData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 },
            ],
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 },
            ],
          },
        ],
        loading: false,
        isError: false,
      },
      datepicker: {
        startDate: '',
        endDate: '',
      },
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Day'));
    expect(container).toMatchSnapshot();
  });
  it('get data success and group by week', () => {
    mockState = {
      lineChart: {
        deviceData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 },
            ],
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 },
            ],
          },
        ],
        loading: false,
        isError: false,
      },
      datepicker: {
        startDate: '',
        endDate: '',
      },
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Week'));
    expect(container).toMatchSnapshot();
  });
  it('get data success and group by month', () => {
    mockState = {
      lineChart: {
        deviceData: [
          {
            name: 'Android',
            data: [
              { x: '01/01/2021', y: 5 },
              { x: '02/01/2021', y: 10 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 4 },
              { x: '02/02/2021', y: 6 },
            ],
          },
          {
            name: 'iOS',
            data: [
              { x: '01/01/2021', y: 2 },
              { x: '02/01/2021', y: 3 },
              { x: '03/01/2021', y: 5 },
              { x: '01/02/2021', y: 10 },
              { x: '02/02/2021', y: 5 },
            ],
          },
        ],
        loading: false,
        isError: false,
      },
      datepicker: {
        startDate: '',
        endDate: '',
      },
    };
    const { container, getByTestId } = render(<LineChart />);
    fireEvent.click(getByTestId('Month'));
    expect(container).toMatchSnapshot();
  });
  it('get data error', () => {
    mockState = {
      lineChart: {
        deviceData: [],
        loading: false,
        isError: true,
      },
      datepicker: {
        startDate: '01/01/2021',
        endDate: '01/02/2021',
      },
    };
    const { container, getByText } = render(<LineChart />);
    expect(container).toMatchSnapshot();
    expect(getByText('Error: Network Error')).toBeTruthy();
  });
});
