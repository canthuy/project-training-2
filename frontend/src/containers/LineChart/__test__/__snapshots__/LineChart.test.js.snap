// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart component get data error 1`] = `
<div>
  <div
    class="sc-dkPtRN bShGKG"
  >
    <h2
      class="sc-gsDKAQ ioZYxa"
    >
      Error: Network Error
    </h2>
  </div>
</div>
`;

exports[`LineChart component get data success 1`] = `
<div>
  <div
    class="sc-dkPtRN bShGKG"
  >
    <div
      class="data-group"
    >
      <h5
        class="sc-hKwDye hRiMkt"
      >
        Device
      </h5>
      <div>
        <span
          data-testid="Day"
          style="opacity: 1;"
        >
          Day
        </span>
        <span
          data-testid="Week"
          style="opacity: 0.3;"
        >
          Week
        </span>
        <span
          data-testid="Month"
          style="opacity: 0.3;"
        >
          Month
        </span>
      </div>
    </div>
    <chart
      height="404"
      options="[object Object]"
      series="[object Object],[object Object]"
      type="line"
    />
  </div>
</div>
`;

exports[`LineChart component loading data 1`] = `
<div>
  <div
    class="sc-dkPtRN bShGKG"
  >
    <div
      class="sc-bdvvtL heGnFz"
    >
      <div
        class="spinner-border"
      />
    </div>
  </div>
</div>
`;
